py/scheduler: Refactor mp_handle_pending to use flags.

iabdalkader opened the pull request:
### Summary

Calling `mp_handle_pending(false)` clears pending exceptions/vm_abort even though they were not handled. This causes pending exceptions/vm_abort to be lost which can be a problem if a tool triggers a vm_abort and waits for soft-reboot...
This patch keeps the pending exceptions and/or VM abort if they were not handled, to allow a following `mp_event_handle_nowait` for example, to deal with them. This seems to me like the correct behavior Vs. clearing pending exceptions, but I could be missing something very very obvious that explains why the current behavior the way it is. If this is the case, what should do then if some code needs to handle pending events but it can't afford raising exceptions, and also can't drop VM aborts?

### Testing

I only did some very limited testing just to confirm that this was indeed what's causing VM aborts to be lost.

### Trade-offs and Alternatives

I guess we could just add our own helper function that restores pending exceptions? For example:
```C
void mp_handle_pending_restore(void) {
    bool vm_abort = MP_STATE_VM(vm_abort);
    mp_obj_t exc_obj = MP_STATE_THREAD(mp_pending_exception);

    mp_handle_pending(false);

    MP_STATE_VM(vm_abort) = vm_abort;
    MP_STATE_THREAD(mp_pending_exception) = exc_obj;
}
```
But I kind of want to avoid leaking VM internals outside of `scheduler.c` and to understand if this break anything.

iabdalkader commented:
> mp_handle_pending(false); // clear any pending exceptions (and run any callbacks)

Thanks for pointing that out. That explains the current behavior of `mp_handle_pending(false);` which seems to be tailored for `pyexec.c`. We have a very different use case that uses `mp_sched_vm_abort` to interrupt the VM/break from REPL, from code that acts like a supervisor. So we can't afford to drop VM abort, yet we also can't let `mp_handle_pending` raise from certain parts of the code... Perhaps splitting that function into two would make more sense, something like:

```C
mp_clear_pending();
mp_handle_pending();
```

dlech reviewed (commented):
I think this should all be done in 1 commit so that it doesn't cause errors in a `git bisect`.

dpgeorge commented:
I agree with @dlech above that the flags are confusing, for two reasons:
- the naming seems off to me, eg "ALL" is not all flags and negative names get confusing
- there are 8 combinations of or'ing flags, but only 3 are really going to be used (see my original suggestion above)

I suggest not using or'able flags and just have the 3 possibilities as enums.  That limits the scope of things, and prevents callers of the function from doing things we didn't think of.  Because this is at the core of the scheduler, it's easy for subtle things to go wrong, and it'd be good to limit the scope for that reason, ie don't allow arbitrary combinations of flags.

> Note: many files call mp_handle_pending(true);, I wonder if those should be calling nowait() instead.

Yes, a lot of ports still need to switch to the new scheme for waiting for events.  Quite a few of the calls to `mp_handle_pending()` will then disappear.

Actually, from that point of view maybe it's simpler to just do as you suggest above: 

> if you would like to minimize the changes, we could probably add a new function and keep the backward compatible function and make it just call with flags based on true/false until it gets deprecated

If you take that approach then we can defer the decision about enum/flags/naming until later.

iabdalkader committed:
py/scheduler: Allow selective handling in mp_handle_pending.

Extend mp_handle_pending to support three distinct behaviors via
mp_handle_pending_internal():

- MP_HANDLE_PENDING_CALLBACKS_ONLY: process callbacks only
- MP_HANDLE_PENDING_CALLBACKS_AND_EXCEPTIONS: callbacks + raise exceptions
- MP_HANDLE_PENDING_CALLBACKS_AND_CLEAR_EXCEPTIONS: callbacks + clear only

Original mp_handle_pending(bool) preserved as inline wrapper for
backward compatibility.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

iabdalkader committed:
stm32/usbd_hid_interface: Add runtime header.

Include runtime.h for mp_handle_pending.

Signed-off-by: iabdalkader <i.abdalkader@gmail.com>

dlech commented:
LGTM

iabdalkader commented on py/runtime.h:
```diff
@@ -100,7 +100,20 @@ void mp_sched_keyboard_interrupt(void);
+    MP_HANDLE_PENDING_CALLBACKS_ONLY,
+    MP_HANDLE_PENDING_CALLBACKS_AND_EXCEPTIONS,
+    MP_HANDLE_PENDING_CALLBACKS_AND_CLEAR_EXCEPTIONS,
+} mp_handle_pending_behaviour_t;
```
Moved.

dpgeorge commented:
Thanks for updating.  Now merged.
